ir/src/constraints/graph.rs

```diff
@@ -58,130 +55,271 @@ impl AlgebraicGraph {
         }
     }
 
-    /// Recursively accumulates the base degree and the cycle lengths of the periodic columns.
-    fn accumulate_degree(&self, cycles: &mut BTreeMap<usize, usize>, index: &NodeIndex) -> usize {
-        // recursively walk the subgraph and compute the degree from the operation and child nodes
-        match self.node(index).op() {
-            Operation::Constant(_) | Operation::RandomValue(_) => 0,
-            Operation::TraceElement(_) => 1,
-            Operation::PeriodicColumn(index, cycle_len) => {
-                cycles.insert(*index, *cycle_len);
-                0
-            }
-            Operation::Neg(index) => self.accumulate_degree(cycles, index),
-            Operation::Add(lhs, rhs) => {
-                let lhs_base = self.accumulate_degree(cycles, lhs);
-                let rhs_base = self.accumulate_degree(cycles, rhs);
-                lhs_base.max(rhs_base)
-            }
+    /// Given a node index that is expected to be the root index of a boundary constraint, returns
+    /// the [IndexedTraceAccess] representing the trace segment and column against which the
+    /// boundary constraint expression must hold, as well as the node index that represents the root
+    /// of the constraint expression that must equal zero during evaluation.
+    pub fn split_boundary_constraint(
+        &self,
+        index: &NodeIndex,
+    ) -> Result<(&IndexedTraceAccess, &NodeIndex), SemanticError> {
```

Comment:
Seems like this is used only in Winterfell codegen, and I'm not sure if this will be relevant in other codgens (or if it will be relevant in the same way). Maybe the logic should go into Winterfell codegen rather than being here?

Reply:
Yeah, I think you're right that it's better there. I've moved it.